torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,392 @@
+    // go down the route where we treat the inputs as tensors
+    // and fully optimize
+    auto & implementation = getOrCompile(inputs);
+    return implementation.run(std::move(inputs));
+  }
+
+private:
+
+  static bool needsGradient(const variable_tensor_list & inputs) {
+    if (!autograd::GradMode::is_enabled()) {
+      return false;
+    }
+    for (const auto & tensor : inputs) {
+      if(tensor.defined() && static_cast<const Variable&>(tensor).requires_grad())
+        return true;
+    }
+    return false;
+  }
+  static bool isDifferentiable(Graph & g) {
+    for(auto n : g.nodes()) {
+      if(!jit::isDifferentiable(n))
+        return false;
+    }
+    return true;
+  }
+
+  void runOptimization(std::shared_ptr<Graph> & graph, bool graphMustSupportVariables) {
+
+    // these optimizations must run in the presence of variables
+    // and when shape information is not statically known.
+
+    EliminateDeadCode(graph);
+    CheckInplace(graph);
+    EliminateCommonSubexpression(graph);
+
+    if (!graphMustSupportVariables) {
+      // These optimizations can introduce operators like FusionGroup that
+      // do not work on variables
+
+      // They also may assume that concrete sizes/strides are availiable
+
+      //TODO: create peephole optimizations that are safe to run
+      // when we are using variables, and when we do not know sizes.
+      PeepholeOptimize(graph);
+      // TODO: remove mandatory size checking in BatchMM, otherwise
+      // it works fine on variables.
+      BatchMM(graph);
+      FuseGraph(graph);
+    }
+  }
+  Code & getOrCreateAutogradFallback() {
+    std::lock_guard<std::mutex> lock(compile_mutex);
+    if(autograd_fallback) {
+      return autograd_fallback;
+    }
+    auto graph_ = graph->copy();
+    if(optimize) {
+      CreateAutodiffSubgraphs(*graph_);
+      runOptimization(graph_, /*graphMustSupportVariables=*/true);
+    }
+    autograd_fallback = Code(graph_);
+    return autograd_fallback;
+  }
+  ExecutionPlan & getOrCompile(const variable_tensor_list & inputs) {
+    // outside lock guard, to minimize the time holding the lock on the fast path
+    // ArgumentSpec even computes its hashCode here.
+    ArgumentSpec spec(autograd::GradMode::is_enabled(), inputs);
+    {
+      std::lock_guard<std::mutex> lock(compile_mutex);
+      auto it = plan_cache.find(spec);
+      if(it != plan_cache.end())
+        return it->second;
+      auto plan = compileSpec(spec);
+      auto r = plan_cache.emplace(std::move(spec), std::move(plan));
+      return r.first->second;
+    }
+  }
+  bool needsGradient(const ArgumentSpec & spec) {
+    for(size_t i = 0; i < spec.size(); ++i) {
+      if(spec.tensorInfo(i).requires_grad())
+        return true;
+    }
+    return false;
+  }
+
+
+  // remove ReplaceIfUndef(v, replacement) nodes that consume inputs with 'v' if
+  // the input is defined, and 'replacement' if it is not.
+  void specializeUndef(Graph & g, const ArgumentSpec & spec) {
```

Comment:
I guess it would be good to point out that explicitly that this is a very limited pass; e.g., if I had `ReplaceIfUndef(Undefined(), replacement)` it wouldn't actually optimize it out. It is *solely* for inputs, and it does *not* propagate newly learned undefinedness facts to the rest of the graph.